### PR TITLE
Forbid label updates by nodes through pod/status

### PR DESCRIPTION
**Pull request by tallclair**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Pod labels can no longer be updated through the pod/status updates by nodes.

**Which issue(s) this PR fixes**:
Part of https://github.com/kubernetes/enhancements/issues/1314

**Special notes for your reviewer**:
Note that this change is NOT feature gated.

**Does this PR introduce a user-facing change?**:
```release-note
Pod labels can no longer be updated through the pod/status updates by nodes.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190916-noderestriction-pods.md
```

/sig auth
/sig node
/milestone v1.17
/priority important-longterm

/assign @liggitt @deads2k

### CODE REVIEW

**Review by mattjmcnaughton — APPROVED** [review]
/lgtm

I saw the original KEP also mentioned restricting the ability to modify annotations. Will that come in a separate diff? Is it a separate diff bc restricting the ability to modify annotations requires a feature gate?

**Comment by tallclair** [comment]
Thanks @mattjmcnaughton 

> I saw the original KEP also mentioned restricting the ability to modify annotations. Will that come in a separate diff? Is it a separate diff bc restricting the ability to modify annotations requires a feature gate?

We ended up dropping all the annotation restrictions from the KEP because those were speculative. I think we could still consider preventing annotation updates through pod/status because I don't think it's ever done by the Kubelet.

File: plugin/pkg/admission/noderestriction/admission.go [hunk 1]
```diff
@@ -235,14 +236,21 @@ func (p *Plugin) admitPodStatus(nodeName string, a admission.Attributes) error {
...
 		// only allow a node to update status of a pod bound to itself
-		if pod.Spec.NodeName != nodeName {
+		if oldPod.Spec.NodeName != nodeName {
 			return admission.NewForbidden(a, fmt.Errorf("node %q can only update pod status for pods with spec.nodeName set to itself", nodeName))
 		}
+		newPod, ok := a.GetObject().(*api.Pod)
+		if !ok {
+			return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
+		}
+		if !labels.Equals(oldPod.Labels, newPod.Labels) {
```
Comment (tedyu) [inline plugin/pkg/admission/noderestriction/admission.go]: It seems reflect.DeepEqual should be used here

Comment (tallclair) [reply plugin/pkg/admission/noderestriction/admission.go]: Why? `labels.Equals` is checking semantic equality and is much faster.

Comment (tedyu) [reply plugin/pkg/admission/noderestriction/admission.go]: I was looking at existing code.
e.g. in pkg/controller/daemon/daemon_controller.go , line 579:
```
	labelChanged := !reflect.DeepEqual(curPod.Labels, oldPod.Labels)
```

Comment (tallclair) [reply plugin/pkg/admission/noderestriction/admission.go]: Yeah, I think it's better to change the existing code to use `labels.Equals`, but it doesn't matter that much - both are logically equivalent.

**Comment by markyjackson-taulia** [comment]
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

**Comment by tallclair** [comment]
@markyjackson-taulia yes

**Pull request merged**